src/sage/crypto/key_exchange/all.py

```diff
@@ -0,0 +1,4 @@
+from sage.misc.lazy_import import lazy_import
+
+lazy_import('sage.crypto.key_exchange.diffie_hellman', 'DiffieHellman')
+del lazy_import
```

Comment:
Didn't think of this before, but seems like a good idea!

Reply:
I just copied from the `crypto/public_key` folder for this.

I'm now thinking I may have done this wrong (and maybe `public_key` did too but I think that's out of scope for this PR, `public_key` needs some refactoring I think). I'm going to do some more testing of this.

Reply:
I was doing it wrong. After the commit I just pushed, `DiffieHellman` can now be accessed without the user needing to explicitly import it.

This does raise the question of how we want the schemes to be accessed by the user. Most people probably won't use this feature (although I guess that applies to most things in Sage) so maybe it makes sense to access it as something like `crypto.DiffieHellman`, to avoid polluting the auto-suggest feature in some interfaces, but several of the existing symmetric ciphers are just accessed directly so I've done the same for consistency.

Happy to change it to something like `crypto.DiffieHellman` or `key_exchange.DiffieHellman` if that's preferred though.

Reply:
Right now I have it so, `DiffieHellman` is accessible without import, but `KeyExchangeScheme` requires an import. I don't think an abstract base class needs to be exposed to users without them explicitly importing it.

Reply:
For my PR of a similar nature, I created a `catalog` file, see [here](https://github.com/sagemath/sage/pull/36581/files#diff-81cd27b8fea2497cd1c834dd59f615575a64134d79461fb383ae5337ff095970). (Suggested by Travis, another Sage contributor)

Reply:
Good idea, I've implemented this.

I still didn't include `KeyExchangeScheme` in the catalogue because it's an abstract base class, but the catalogue should be useful as more key exchange schemes are added in the future.

Reply:
I think having `key_exchange` in global scope (I think?) is *a bit* questionable. Do you think it's a good idea?

Reply:
I think it's less questionable than `SubstitutionCryptosystem` being global scope (which it currently is).

I don't think it's too different than how `distributions` is global scope in the PR you linked. The alternative (which maybe is better) would be to access the key exchange schemes under `crypto.key_exchange`, so you'd do `crypto.key_exchange.DiffieHellman`. We could then in the future move the other cryptography code to `crypto.[something]` as well and take things like `SubstitutionCryptosystem` out of global scope.

After writing that I've convinced myself that it should be changed to `crypto.key_exchange`.

And the reason I didn't consider putting it under `crypto` directly (so `crypto.DiffieHellman`) is some schemes (like RSA) have multiple uses. RSA can be used for public key encryption, key exchange, and cryptographic signing. I think having something like `crypto.pke.RSA`, `crypto.key_exchange.RSA`, and `crypto.signing.RSA` would be better than `crypto.RSAEncryption`, `crypto.RSAKeyExchange`, and `crypto.RSASign`. Especially because for discoverability, a user is more likely to want to know "what key exchange schemes are in Sage" than "what RSA variants are in Sage". Thinking about the future and how many cryptography schemes (public key encryption, key exchange, signing, symmetric ciphers, etc.) could be implemented in Sage, putting it all under `crypto` would be too much.

Reply:
I think `crypto.key_exchange.DiffieHellman` looks good (definitely not `crypto.DiffieHellman`), but again I don't have a strong opinion. 

> After writing that I've convinced myself that it should be changed to crypto.key_exchange.

Lol, it happens :)

Reply:
Okay, I realized that making it accessible under `crypto.key_exchange` would involve some major restructuring to the `crypto` module and I think that would be out of scope for this PR. Unless I'm forgetting some syntax, I don't think Python import statements let you do `from a.b.c import x as y.z`. Basically above I was advocating to change `from sage.crypto.all import *` to `import sage.crypto.all as crypto` in the `all.py` file in `src/sage` which currently only `stats` does (but `stats` still imports everything to global scope anyway).

For now I think it's fine to leave it as `key_exchange` in global scope. It's at least better than what we have for other stuff in the `crypto` module. But if you have other suggestions I'm open.

Reply:
Okay if it's too much irrelevant stuff for this PR then it's okay to leave it (and optionally make a new issue that hopefully someone gets to.)